hub/saml-engine/src/test/java/uk/gov/ida/hub/samlengine/metadata/SigningCertFromMetadataExtractorTest.java

```diff
@@ -0,0 +1,103 @@
+import org.opensaml.saml.metadata.resolver.MetadataResolver;
+import org.opensaml.saml.saml2.metadata.EntityDescriptor;
+import org.opensaml.saml.saml2.metadata.KeyDescriptor;
+import org.opensaml.security.credential.UsageType;
+import org.opensaml.xmlsec.signature.support.SignatureException;
+import uk.gov.ida.common.shared.security.X509CertificateFactory;
+import uk.gov.ida.hub.samlengine.exceptions.CertificateForCurrentPrivateSigningKeyNotFoundInMetadataException;
+import uk.gov.ida.hub.samlengine.exceptions.UnableToResolveSigningCertsForHubException;
+import uk.gov.ida.saml.core.test.TestCertificateStrings;
+import uk.gov.ida.saml.core.test.TestEntityIds;
+import uk.gov.ida.saml.core.test.builders.metadata.EntityDescriptorBuilder;
+import uk.gov.ida.saml.core.test.builders.metadata.SPSSODescriptorBuilder;
+
+import java.security.PublicKey;
+import java.security.cert.X509Certificate;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.mockito.Matchers.any;
+import static org.mockito.Mockito.when;
+import static uk.gov.ida.saml.core.test.TestEntityIds.HUB_ENTITY_ID;
+import static uk.gov.ida.saml.core.test.TestEntityIds.TEST_RP;
+import static uk.gov.ida.saml.core.test.builders.metadata.KeyDescriptorBuilder.aKeyDescriptor;
+import static uk.gov.ida.saml.core.test.builders.metadata.KeyInfoBuilder.aKeyInfo;
+import static uk.gov.ida.saml.core.test.builders.metadata.X509CertificateBuilder.aX509Certificate;
+import static uk.gov.ida.saml.core.test.builders.metadata.X509DataBuilder.aX509Data;
+
+@RunWith(MockitoJUnitRunner.class)
+public class SigningCertFromMetadataExtractorTest {
+
+    private SigningCertFromMetadataExtractor signingCertFromMetadataExtractor;
+
+    private static EntityDescriptor hubEntityDescriptor;
+
+    @BeforeClass
+    public static void beforeClass() throws MarshallingException, SignatureException {
+        KeyDescriptor secondKeyDescriptor = aKeyDescriptor().withKeyInfo(aKeyInfo().withKeyName(TestEntityIds.HUB_ENTITY_ID).withX509Data(aX509Data().withX509Certificate(aX509Certificate().withCert(TestCertificateStrings.HUB_TEST_SECONDARY_PUBLIC_SIGNING_CERT).build()).build()).build()).withUse(UsageType.SIGNING.toString()).build();
+        hubEntityDescriptor = EntityDescriptorBuilder.anEntityDescriptor()
+                .withEntityId(HUB_ENTITY_ID)
+                .addSpServiceDescriptor(SPSSODescriptorBuilder.anSpServiceDescriptor()
+                        .addKeyDescriptor(secondKeyDescriptor)
+                        .build())
+                .build();
+    }
+
+    private X509Certificate hubPrimarySigningCert = new X509CertificateFactory().createCertificate(TestCertificateStrings.HUB_TEST_PUBLIC_SIGNING_CERT);
+    private X509Certificate hubSecondarySigningCert = new X509CertificateFactory().createCertificate(TestCertificateStrings.HUB_TEST_SECONDARY_PUBLIC_SIGNING_CERT);
+
+    //Intentionally deriving these from a different object instance.
```

Comment:
I don't understand this comment

Reply:
I wanted to make sure the public keys we make the call with are not the same object instances that are in the certificates that we supply in the metadata. In the extractor we check that the public key in the cert we get back is the same as the public key we request with, but outside a test like this they will not be the same object instance. So making that the case here makes the test more lifelike. I wanted the comment to highlight that we were doing this intentionally as otherwise if you need the public key you can just get it out of the cert we create a couple of lines above rather than creating the cert object again.

Reply:
That makes sense. I think the comment could be clearer though - perhaps `Intentionally deriving these PublicKeys from new Certificate instances so they don't have the same object identity as elsewhere` ?